engines/freescape/gfx.h

```diff
@@ -0,0 +1,166 @@
+	Graphics::PixelFormat _palettePixelFormat;
+	Graphics::PixelFormat _texturePixelFormat;
+	bool _isAccelerated;
+
+	virtual void init() = 0;
+	virtual void clear() = 0;
+	virtual void setViewport(const Common::Rect &rect) = 0;
+
+	/**
+	 *  Swap the buffers, making the drawn screen visible
+	 */
+	virtual void flipBuffer() {}
+	virtual void useColor(uint8 r, uint8 g, uint8 b) = 0;
+	virtual void polygonOffset(bool enabled) = 0;
+
+	virtual Texture *createTexture(const Graphics::Surface *surface) = 0;
+	void convertImageFormatIfNecessary(Graphics::Surface *surface);
+
+	virtual void freeTexture(Texture *texture) = 0;
+	virtual void drawTexturedRect2D(const Common::Rect &screenRect, const Common::Rect &textureRect, Texture *texture) = 0;
+
+	virtual void renderCrossair(byte color, const Common::Point position) = 0;
+	virtual void renderShoot(byte color, const Common::Point position) = 0;
+	virtual void renderCube(const Math::Vector3d &position, const Math::Vector3d &size, Common::Array<uint8> *colours);
+	virtual void renderRectangle(const Math::Vector3d &position, const Math::Vector3d &size, Common::Array<uint8> *colours);
+	virtual void renderPolygon(const Math::Vector3d &origin, const Math::Vector3d &size, const Common::Array<uint16> *ordinates, Common::Array<uint8> *colours);
+	virtual void renderPyramid(const Math::Vector3d &origin, const Math::Vector3d &size, const Common::Array<uint16> *ordinates, Common::Array<uint8> *colours, int type);
+	virtual void renderFace(const Common::Array<Math::Vector3d> &vertices) = 0;
+
+	virtual void setSkyColor(uint8 color) = 0;
+	virtual void drawFloor(uint8 color) = 0;
+
+	Common::Rect viewport() const;
+
+	// palette
+	void readFromPalette(uint8 index, uint8 &r, uint8 &g, uint8 &b);
+	bool getRGBAt(uint8 index, uint8 &r, uint8 &g, uint8 &b);
+	byte *_palette;
+	ColorMap *_colorMap;
+	int _keyColor;
+
+	/**
+	 * Select the window where to render
+	 *
+	 * This also sets the viewport
+	 */
+
+	virtual void positionCamera(const Math::Vector3d &pos, const Math::Vector3d &interest) = 0;
+	virtual void updateProjectionMatrix(float fov, float nearClipPlane, float farClipPlane) = 0;
+
+	Math::Matrix4 getMvpMatrix() const { return _mvpMatrix; }
+
+	void flipVertical(Graphics::Surface *s);
+
+	int _screenW;
+	int _screenH;
+	Common::RenderMode _renderMode;
+
+	bool computeScreenViewport();
+
+protected:
+	OSystem *_system;
+	Common::Rect _screenViewport;
+	Common::Rect _viewport;
+	Common::Rect _unscaledViewport;
+
+	Math::Matrix4 _projectionMatrix;
+	Math::Matrix4 _modelViewMatrix;
+	Math::Matrix4 _mvpMatrix;
+
+	Math::Frustum _frustum;
+
+	Math::Matrix4 makeProjectionMatrix(float fov, float nearClipPlane, float farClipPlane) const;
+};
+
+/**
+ * A framerate limiter
+ *
+ * Ensures the framerate does not exceed the specified value
+ * by delaying until all of the timeslot allocated to the frame
+ * is consumed.
+ * Allows to curb CPU usage and have a stable framerate.
+ */
+class FrameLimiter {
```

Comment:
why FrameLimiter here? we have implementation in common graphics

Reply:
It was actually unused. Removed.